bowling/Game.cpp

```diff
@@ -1,203 +1,140 @@
+                players.emplace_back(PlayerData(playerName, playerRolls));
+            }
         }
     }
     file.close();
 }
 
-void Game::loadPlayerRolls(std::ifstream& file, std::vector<Frame>& playerRolls) {
-    std::string rolls{};
+std::vector<Frame> Game::getPlayerRolls(std::ifstream& file) {
+    std::vector<Frame> playerRolls{};
+    std::string line{};
+    std::getline(file, line);
+    playerRolls = getFrames(line);
+    return playerRolls;
+}
+
+std::vector<Frame> Game::getFrames(const std::string& line) {
+    std::vector<Frame> playerRolls{};
     Frame currentFrame{};
-    int frameCount = 1;
-    std::getline(file, rolls);
-    for (int currentRoll = 0; currentRoll < rolls.size(); ++currentRoll) {
-        if (isStrike(rolls[currentRoll])) {
+    for (int currentChar = 0, frameNumber = 1; currentChar < line.size(); ++frameNumber) {
+        if (isStrike(line[currentChar])) {
             currentFrame = Frame('X', ' ');
-            currentRoll++;
-            if (isLastFrame(frameCount)) {
-                currentRoll++;
-            }
-        } else if (isSpare(rolls[currentRoll + 1])) {
-            currentFrame = Frame(rolls[currentRoll], '/');
-            currentRoll += 2;
-            if (isLastFrame(frameCount)) {
-                currentRoll++;
-            }
+            incrementCharForStrike(currentChar, frameNumber);
+        } else if (isSpare(line[currentChar + 1])) {
+            currentFrame = Frame(line[currentChar], '/');
+            incrementCharForSpare(currentChar, frameNumber);
         } else {
-            if (currentRoll == rolls.size() - 1) {
-                currentFrame = Frame(rolls[currentRoll], ' ');
-            } else {
-                currentFrame = Frame(rolls[currentRoll], rolls[currentRoll + 1]);
-            }
-            currentRoll += 2;
+            currentFrame = getRegularFrame(line, currentChar);
+            currentChar += 3;
         }
-        frameCount++;
         playerRolls.push_back(currentFrame);
     }
+    return playerRolls;
 }
 
-bool Game::isLastFrame(int frameCount) const {
-    return frameCount == 10;
+Frame Game::getRegularFrame(const std::string& rolls, int currentRoll) const {
+    if (currentRoll == rolls.size() - 1) {
+        return Frame(rolls[currentRoll], ' ');
+    }
+    return Frame(rolls[currentRoll], rolls[currentRoll + 1]);
 }
 
-bool Game::isStrike(char currentRoll) const {
-    return currentRoll == 'X';
+void Game::incrementCharForStrike(int& currentRoll, int frameNumber) const {
+    currentRoll += 2;
+    if (isLastFrame(frameNumber)) {
+        currentRoll++;
+    }
+}
+
+void Game::incrementCharForSpare(int& currentRoll, int frameNumber) const {
+    currentRoll += 3;
+    if (isLastFrame(frameNumber)) {
+        currentRoll++;
+    }
 }
 
-bool Game::isStrike(const Frame& frame) const {
-    return frame.getFirstRoll() == 'X' ? true : false;
+bool Game::isLastFrame(int frameNumber) const {
+    return frameNumber == 10;
```

Comment:
what about bonus balls?

Reply:
This is a supplementary function used to move the cursor behind additional '|' in order to count points for bonus balls. Perhaps the name is a bit misleading..